File: src/main/kotlin/io/titandata/titan/commands/Pull.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.commands

import io.titandata.titan.Dependencies
import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.requireObject
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.multiple
import com.github.ajalt.clikt.parameters.options.option
import org.kodein.di.Kodein
import org.kodein.di.generic.bind
import org.kodein.di.generic.inSet
import org.kodein.di.generic.provider

class Pull : CliktCommand(help = "Pull a new data state from remote") {
    private val dependencies: Dependencies by requireObject()
    private val commit by option("-c", "--commit", help="Commit GUID to pull from, defaults to latest")
    private val remote: String? by option("-r", "--remote", help="Name of the remote provider, defaults to origin")
    private val metadataOnly: Boolean by option("-u", "--update-only", help="Update tags only, do not pull data").flag()
    private val tags by option("-t", "--tag", help="Filter commits to select commit to pull").multiple()
    private val repository by argument()

    override fun run() {
        val provider = dependencies.providers.byRepository(repository)
        provider.pull(repository, commit, remote, tags, metadataOnly)
    }
}

val pullModule = Kodein.Module("pull") {
    bind<CliktCommand>().inSet() with provider {
        Pull()
    }
}

File: internal/app/commands/pull.go
/*
Copyright © 2019 The Titan Project Contributors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package commands

import (
	"github.com/spf13/cobra"
)


// pullCmd represents the pull command
var pullCmd = &cobra.Command{
	Use:   "pull [REPOSITORY]",
	Short: "Pull a new data state from remote",
	Args: cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		repo := args[0]
		provider.Pull(repo, guid, remote, tags, updateOnly)
	},
}

func init() {
	rootCmd.AddCommand(pullCmd)
	pullCmd.Flags().StringVarP(&guid, "commit","c", "","commit GUID to pull from, defaults to latest")
	pullCmd.Flags().StringVarP(&remote, "remote","r", "","name of the remote provider, defaults to origin")
	pullCmd.Flags().BoolVarP(&updateOnly, "update-only", "u", false, "update tags only, do not pull data")
	pullCmd.Flags().StringSliceVarP(&tags, "tags", "t", nil, "filter commits to select commit to pull")
	pullCmd.Flags().SortFlags = false //TODO review flag sorting
}
